### PR TITLE
Add http_sd_configs

### PR DESCRIPTION
**Pull request by loafoe**
This config option was added in Prometheus `2.28.0` and it would be great if this was supported in this package

### CODE REVIEW

File: http_client.go [hunk 1]
```diff
@@ -18,20 +18,57 @@
...
+	OAuth2 *OAuth2 `yaml:"oauth2,omitempty" json:"oauth2,omitempty"`
+	// The bearer token for the targets. Deprecated in favour of
+	// Authorization.Credentials.
+	BearerToken Secret `yaml:"bearer_token,omitempty" json:"bearer_token,omitempty"`
+	// The bearer token file for the targets. Deprecated in favour of
+	// Authorization.CredentialsFile.
+	BearerTokenFile string `yaml:"bearer_token_file,omitempty" json:"bearer_token_file,omitempty"`
 	// HTTP proxy server to use to connect to the targets.
-	ProxyURL string `yaml:"proxy_url,omitempty"`
+	ProxyURL URL `yaml:"proxy_url,omitempty" json:"proxy_url,omitempty"`
```
Comment (palash25) [inline http_client.go]: Hi sorry for the late reply and thanks for your contribution.

we use these fields in pmm-managed while configuring alertmanager, I will have to check whether merging this is desirable and how many changes will it require for us to do on pmm-managed side.

Will get back to this later today.

**Review by palash25 — APPROVED** [review]
sorry for another late response, LGTM

the only thing we will have to change on pmm-managed after merging this is https://github.com/percona/pmm-managed/blob/PMM-2.0/services/alertmanager/alertmanager.go#L531  where it accepts the ProxyURL to be a plain string.


I think we can merge this, thank you for your contribution

File: http_client.go [hunk 2]
```diff
@@ -18,20 +18,57 @@
...
+
 // HTTPClientConfig configures an HTTP client.
 type HTTPClientConfig struct {
 	// The HTTP basic authentication credentials for the targets.
-	BasicAuth *BasicAuth `yaml:"basic_auth,omitempty"`
-	// The bearer token for the targets.
-	BearerToken string `yaml:"bearer_token,omitempty"`
-	// The bearer token file for the targets.
-	BearerTokenFile string `yaml:"bearer_token_file,omitempty"`
+	BasicAuth *BasicAuth `yaml:"basic_auth,omitempty" json:"basic_auth,omitempty"`
```
Comment (BupycHuk) [inline http_client.go]: did you add `json` here for a reason?
It adds changes in our test which is unexpected.

File: http_client.go [hunk 3]
```diff
@@ -18,20 +18,57 @@
...
 	// TLSConfig to use to connect to the targets.
-	TLSConfig TLSConfig `yaml:"tls_config,omitempty"`
+	TLSConfig TLSConfig `yaml:"tls_config,omitempty" json:"tls_config,omitempty"`
+	// FollowRedirects specifies whether the client should follow HTTP 3xx redirects.
+	// The omitempty flag is not set, because it would be hidden from the
+	// marshalled configuration when set to false.
+	FollowRedirects bool `yaml:"follow_redirects" json:"follow_redirects"`
+}
+
+// URL is a custom URL type that allows validation at configuration load time.
```
Comment (BupycHuk) [inline http_client.go]: i see a comment, but I don't see validation

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Could you add new test cases, please?

File: http_client.go (hunk 2 above)
Comment (artemgavrilov) [reply http_client.go]: Also `json` should be added either everywhere or nowhere.

File: http_client.go [hunk 4]
```diff
@@ -18,20 +18,57 @@
...
+	ClientSecret     Secret            `yaml:"client_secret" json:"client_secret"`
+	ClientSecretFile string            `yaml:"client_secret_file" json:"client_secret_file"`
+	Scopes           []string          `yaml:"scopes,omitempty" json:"scopes,omitempty"`
+	TokenURL         string            `yaml:"token_url" json:"token_url"`
+	EndpointParams   map[string]string `yaml:"endpoint_params,omitempty" json:"endpoint_params,omitempty"`
 }
 
+// Secret special type for storing secrets.
+type Secret string
+
```
Comment (artemgavrilov) [inline http_client.go]: I think we don't need this special type. Prometheus uses it to hide the actual value of the secret when they marshall config, but it's not what we need in PMM. 

https://github.com/prometheus/common/blob/8d1c9f84e3f78cb628a20f8e7be531c508237848/config/config.go#L24-L49

File: http_client.go (hunk 3 above)
Comment (artemgavrilov) [reply http_client.go]: I think here is the missing part: https://github.com/prometheus/common/blob/8d1c9f84e3f78cb628a20f8e7be531c508237848/config/http_config.go#L96-L116

**Review by artemgavrilov — CHANGES_REQUESTED** [review]
Thanks for your contribution!

File: http_client.go (hunk 3 above)
Comment (BupycHuk) [reply http_client.go]: And this part too, to have less changes in our test case https://github.com/prometheus/common/blob/8d1c9f84e3f78cb628a20f8e7be531c508237848/config/http_config.go#L118-L138

File: http_client.go (hunk 2 above)
Comment (loafoe) [reply http_client.go]: The `json` just came as part of the copy-paste I did from Prometheus structs. I will remove the tags as I didn't realize this project stripped these. Will also fix the tests.

File: http_client.go (hunk 3 above)
Comment (loafoe) [reply http_client.go]: Will add the missing YAML marshaller methods

File: http_client.go (hunk 4 above)
Comment (loafoe) [reply http_client.go]: Removed the Secret type

File: http_client.go (hunk 1 above)
Comment (loafoe) [reply http_client.go]: I think this can be reverted back to `string` so as to not affect any dependencies

**Review by artemgavrilov — APPROVED** [review]

**Review by BupycHuk — APPROVED** [review]

**Pull request merged**